### DISCUSSION TITLE
I don't understand the pruning examples

### DISCUSSION THREAD

**Discussion by tvdboom** [optuna/optuna#3984] [Q&A] [answered]
Hi. I am trying to understand the examples used when explaining how pruning works. For example this one [here](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.pruners.PatientPruner.html#optuna.pruners.PatientPruner):

```python
import numpy as np
from sklearn.datasets import load_iris
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import train_test_split

import optuna

X, y = load_iris(return_X_y=True)
X_train, X_valid, y_train, y_valid = train_test_split(X, y)
classes = np.unique(y)


def objective(trial):
    alpha = trial.suggest_float("alpha", 0.0, 1.0)
    clf = SGDClassifier(alpha=alpha)
    n_train_iter = 100

    for step in range(n_train_iter):
        clf.partial_fit(X_train, y_train, classes=classes)

        intermediate_value = clf.score(X_valid, y_valid)
        trial.report(intermediate_value, step)

        if trial.should_prune():
            raise optuna.TrialPruned()

    return clf.score(X_valid, y_valid)


study = optuna.create_study(
    direction="maximize",
    pruner=optuna.pruners.PatientPruner(optuna.pruners.MedianPruner(), patience=1),
)
study.optimize(objective, n_trials=20)
```

The way I understand it, `partial_fit` should be used for batch learning, as explained [here](https://scikit-learn.org/0.15/modules/scaling_strategies.html#incremental-learning). The data provided to `partial_fit` should be a batch (or part) of the complete dataset. In this example however (and the other ones), `partial_fit` is used to fit SGDClassifier 100 times on the exact same dataset (same rows, not even shuffled). I don't think this is how `partial_fit` is supposed to be used. What am I missing? Any help appreciated.

**Comment by nzw0301** [comment, marked as answer]
> I don't think this is how `partial_fit` is supposed to be used.

No. As described in https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier.partial_fit, single call of `partial_fit` is equivalent to `fit` method with `max_iter=1`, so it is totally fine with using the full dataset as in the Optuna example. Of course, if you prefer to use a stochastic way like a stochastic gradient descent algorithm to train large neural networks on a large dataset, you can use a part of dataset as an argument of `partial_fit`.

**Reply by tvdboom** [reply]
I see. I guess in this case `partial_fit` is slightly different than for other estimators like [here](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html#sklearn.naive_bayes.MultinomialNB.partial_fit) for example. Thanks for the explanation!
